File: backend/app/core/season.py
"""Helpers de saison sportive : du 1ᵉʳ septembre Y au 31 août Y+1.

Module pur (aucune dépendance DB). L'identifiant d'une saison est son année de
début Y. La saison Y couvre [Y-09-01, (Y+1)-08-31] et s'affiche « Saison Y — Y+1 ».
"""
from datetime import date

from app.core.time import utcnow


def season_of(d: date) -> int:
    """Année de début de la saison contenant `d` (bascule au 1ᵉʳ septembre)."""
    return d.year if d.month >= 9 else d.year - 1


def season_bounds(start_year: int) -> tuple[date, date]:
    """Bornes incluses (date_from, date_to) de la saison d'année de début `start_year`."""
    return date(start_year, 9, 1), date(start_year + 1, 8, 31)


def current_season() -> int:
    """Saison en cours, calculée depuis l'horloge centralisée (figeable en test)."""
    return season_of(utcnow().date())


def season_label(start_year: int) -> str:
    """Libellé d'affichage « Saison Y — Y+1 »."""
    return f"Saison {start_year} — {start_year + 1}"


def parse_seasons(raw: str | None) -> list[int]:
    """Parse un CSV d'années de début (« 2025,2023 ») → liste d'entiers.

    Tolère les espaces, ignore les valeurs non entières, dédoublonne en
    conservant l'ordre d'apparition.
    """
    if not raw:
        return []
    out: list[int] = []
    for token in raw.split(","):
        token = token.strip()
        if not token:
            continue
        try:
            year = int(token)
        except ValueError:
            continue
        if year not in out:
            out.append(year)
    return out

File: frontend/lib/utils/season.ts
// Miroir des helpers backend `app/core/season.py`. Saison = 1ᵉʳ sept Y → 31 août Y+1.
// Duplication assumée (le front ne partage pas de code Python) ; couverte par tests de bornes.

/** Année de début de la saison contenant la date ISO « YYYY-MM-DD ». */
export function seasonOf(iso: string): number {
  const year = Number(iso.slice(0, 4));
  const month = Number(iso.slice(5, 7));
  return month >= 9 ? year : year - 1;
}

/** Saison en cours (bascule au 1ᵉʳ septembre). `now` injectable pour les tests.
 *  Getters UTC pour rester miroir du backend (`utcnow()`) et éviter tout
 *  mismatch SSR/hydratation autour du 1ᵉʳ septembre selon le fuseau local. */
export function currentSeason(now: Date = new Date()): number {
  const year = now.getUTCFullYear();
  const month = now.getUTCMonth() + 1;
  return month >= 9 ? year : year - 1;
}

/** Libellé d'affichage « Saison Y — Y+1 ». */
export function seasonLabel(startYear: number): string {
  return `Saison ${startYear} — ${startYear + 1}`;
}

/** Parse un CSV d'années (« 2025,2023 ») : tolère espaces, ignore non entiers, dédoublonne. */
export function parseSeasonsParam(raw?: string | null): number[] {
  if (!raw) return [];
  const out: number[] = [];
  for (const token of raw.split(",")) {
    const trimmed = token.trim();
    if (!trimmed) continue;
    const year = Number(trimmed);
    if (!Number.isInteger(year)) continue;
    if (!out.includes(year)) out.push(year);
  }
  return out;
}

/** Sérialise une liste d'années en CSV. */
export function serializeSeasons(years: number[]): string {
  return years.join(",");
}

/** Ajoute/retire une saison de la sélection (toggle). */
export function toggleSeason(selected: number[], year: number): number[] {
  return selected.includes(year)
    ? selected.filter((y) => y !== year)
    : [...selected, year];
}

/** Libellé de l'en-tête : 1 saison → libellé complet ; plusieurs → décompte. */
export function seasonSelectionLabel(years: number[]): string {
  if (years.length === 0) return seasonLabel(currentSeason());
  if (years.length === 1) return seasonLabel(years[0]);
  return `${years.length} saisons sélectionnées`;
}
